src/SAWScript/Yosys/CompositionalTranslation.hs

```diff
@@ -0,0 +1,343 @@
+{-# Language TemplateHaskell #-}
+{-# Language ConstraintKinds #-}
+{-# Language FlexibleContexts #-}
+{-# Language RecordWildCards #-}
+{-# Language OverloadedStrings #-}
+{-# Language LambdaCase #-}
+{-# Language TupleSections #-}
+{-# Language ScopedTypeVariables #-}
+{-# Language ViewPatterns #-}
+
+module SAWScript.Yosys.CompositionalTranslation where
+
+import Control.Lens.TH (makeLenses)
+import Control.Lens ((^.))
+import Control.Monad (forM, (>=>), void)
+import Control.Monad.IO.Class (MonadIO(..))
+import Control.Exception (throw)
+
+import Data.Bifunctor (bimap)
+import qualified Data.Maybe as Maybe
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Data.Map (Map)
+import qualified Data.Map as Map
+
+import qualified Verifier.SAW.SharedTerm as SC
+import qualified Verifier.SAW.Name as SC
+
+import qualified Cryptol.TypeCheck.Type as C
+
+import SAWScript.Panic (panic)
+
+import SAWScript.Yosys.Utils
+import SAWScript.Yosys.IR
+import SAWScript.Yosys.Cell
+
+type ModuleName = Text
+type CellName = Text
+type Pattern = [Bitrep]
+type PatternMap m = Map Pattern ((YosysBitvecConsumer -> Pattern -> m SC.Term) -> m SC.Term)
+
+data CellStateInfo = CellStateInfo
+  { _cellStateInfoType :: SC.Term -- cell state type - either a bitvector for a $dff, or a record type
+  , _cellStateInfoCryptolType :: C.Type -- cryptol type for the above
+  , _cellStateInfoFields :: Maybe (Map Text (SC.Term, C.Type)) -- if the type is a record, the fields of the record
+  }
+makeLenses ''CellStateInfo
+
+data TranslatedModule = TranslatedModule
+  { _translatedModuleStateInfo :: Maybe CellStateInfo -- information about the state type for this module
+  , _translatedModuleTerm :: SC.Term -- the lambda term for the output record (including state) in terms of the inputs (including state)
+  , _translatedModuleType :: SC.Term
+  , _translatedModuleCryptolType :: C.Type
+  }
+makeLenses ''TranslatedModule
+
+data TranslationContext m = TranslationContext
+  { _translationContextModules :: Map ModuleName TranslatedModule
+  , _translationContextStateTypes :: Map CellName CellStateInfo -- state type for every stateful cell in this module (including sequential submodules)
+  , _translationContextPatternMap :: PatternMap m -- for each pattern, a term representing that pattern (parameterized by a function to get a term representing any other pattern)
+  }
+makeLenses ''TranslationContext
+
+buildTranslationContextStateTypes ::
+  MonadIO m =>
+  SC.SharedContext ->
+  Map ModuleName TranslatedModule ->
+  Module ->
+  m (Map CellName CellStateInfo)
+buildTranslationContextStateTypes sc mods m = do
+  fmap (Map.mapMaybe id) . forM (m ^. moduleCells) $ \c -> do
+    case Map.lookup (c ^. cellType) mods of
+      Just tm -> pure $ tm ^. translatedModuleStateInfo
+      Nothing -> case c ^. cellType of
+        "$dff" | Just w <- length <$> Map.lookup "Q" (c ^. cellConnections) -> do
```

Comment:
There are quite a few magic string values that frequently get pattern-matched against when looking up cell types, such as `"$dff"`. (Or, at the very least, they look magical to my untrained eyes.) Do you think it's worth changing `cellType`'s type to be a proper sum type to reflect this?

Reply:
I agree. I added a sum type for cell types in 853a0453c7705055f683919209d08367a283deb8